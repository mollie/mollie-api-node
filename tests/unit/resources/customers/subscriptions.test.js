import axios from 'axios';
import MockAdapter from 'axios-mock-adapter';

import CustomersSubscriptions from 'resources/customers/subscriptions';
import Subscription from 'models/subscription';

import response from '../../__stubs__/customers_subscriptions.json';

const mock = new MockAdapter(axios);

const props = {
  id: 'sub_rVKGtNd6s3',
  customerId: 'cst_stTC2WHAuS',
  amount: {
    currency: 'EUR',
    value: '25.00',
  },
  times: 4,
  interval: '3 months',
  description: 'Quarterly payment',
  method: 'creditcard',
  webhookUrl: 'https://example.org/payments/webhook',
};

describe('customers_subscriptions', () => {
  let customersSubscriptions;
  beforeEach(() => {
    customersSubscriptions = new CustomersSubscriptions(axios.create());
  });

  it('should have a resource name and model', () => {
    expect(CustomersSubscriptions.resource).toBe('customers_subscriptions');
    expect(CustomersSubscriptions.model).toBe(Subscription);
  });

  describe('.create()', () => {
    mock
      .onPost(`/customers/${props.customerId}/subscriptions`)
      .reply(200, response._embedded.subscriptions[0]);

    it('should return a subscription instance', () =>
      customersSubscriptions.create(props).then(result => {
        expect(result).toBeInstanceOf(Subscription);
        expect(result).toMatchSnapshot();
      }));
  });

  describe('.get()', () => {
    const error = {
      error: { message: 'The customers_subscription id is invalid' },
    };

    mock
      .onGet(`/customers/${props.customerId}/subscriptions/${props.id}`)
      .reply(200, response._embedded.subscriptions[0]);
    mock.onGet(`/customers/${props.customerId}/subscriptions/foo`).reply(500, error);

    it('should return a subscription instance', () =>
      customersSubscriptions.get(props.id, { customerId: props.customerId }).then(result => {
        expect(result).toBeInstanceOf(Subscription);
        expect(result).toMatchSnapshot();
      }));
  });

  describe('.all()', () => {
    mock.onGet(`/customers/${props.customerId}/subscriptions`).reply(200, response);

    it('should return a list of all customer subscriptions', () =>
      customersSubscriptions.all({ customerId: props.customerId }).then(result => {
        expect(result).toBeInstanceOf(Array);
        expect(result).toHaveProperty('links');
        expect(result).toMatchSnapshot();
      }));

    it('should work with a callback', done => {
      customersSubscriptions
        .withParent({
          resource: 'customer',
          id: props.customerId,
        })
        .all((err, result) => {
          expect(err).toBeNull();
          expect(result).toBeInstanceOf(Array);
          expect(result).toHaveProperty('links');
          expect(result).toMatchSnapshot();
          done();
        });
    });
  });

  describe('.cancel()', () => {
    mock
      .onDelete(`/customers/${props.customerId}/subscriptions/${props.id}`)
      .reply(200, response._embedded.subscriptions[0]);

    it('should return a subscription instance', () =>
      customersSubscriptions.delete(props.id, { customerId: props.customerId }).then(result => {
        expect(result).toBeInstanceOf(Subscription);
        expect(result).toMatchSnapshot();
      }));

    it('should work with a callback', done => {
      customersSubscriptions
        .withParent({
          resource: 'customer',
          id: props.customerId,
        })
        .cancel(props.id, (err, result) => {
          expect(err).toBeNull();
          expect(result).toBeInstanceOf(Subscription);
          expect(result).toMatchSnapshot();
          done();
        });
    });
  });
});
