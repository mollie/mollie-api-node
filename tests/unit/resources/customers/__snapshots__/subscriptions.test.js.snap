// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`customers_subscriptions .all() should return a list of all customer subscriptions 1`] = `
Array [
  Subscription {
    "_links": Object {
      "customer": Object {
        "href": "https://api.mollie.com/v2/customers/cst_stTC2WHAuS",
        "type": "application/hal+json",
      },
      "self": Object {
        "href": "https://api.mollie.com/v2/customers/cst_stTC2WHAuS/subscriptions/sub_rVKGtNd6s3",
        "type": "application/hal+json",
      },
    },
    "amount": Object {
      "currency": "EUR",
      "value": "25.00",
    },
    "canceledAt": null,
    "createdAt": "2018-06-01T12:23:34+00:00",
    "description": "Quarterly payment",
    "id": "sub_rVKGtNd6s3",
    "interval": "3 months",
    "method": null,
    "mode": "live",
    "resource": "subscription",
    "startDate": null,
    "status": "active",
    "times": 4,
    "webhookUrl": "https://webshop.example.org/subscriptions/webhook",
  },
]
`;

exports[`customers_subscriptions .all() should work with a callback 1`] = `
Array [
  Subscription {
    "_links": Object {
      "customer": Object {
        "href": "https://api.mollie.com/v2/customers/cst_stTC2WHAuS",
        "type": "application/hal+json",
      },
      "self": Object {
        "href": "https://api.mollie.com/v2/customers/cst_stTC2WHAuS/subscriptions/sub_rVKGtNd6s3",
        "type": "application/hal+json",
      },
    },
    "amount": Object {
      "currency": "EUR",
      "value": "25.00",
    },
    "canceledAt": null,
    "createdAt": "2018-06-01T12:23:34+00:00",
    "description": "Quarterly payment",
    "id": "sub_rVKGtNd6s3",
    "interval": "3 months",
    "method": null,
    "mode": "live",
    "resource": "subscription",
    "startDate": null,
    "status": "active",
    "times": 4,
    "webhookUrl": "https://webshop.example.org/subscriptions/webhook",
  },
]
`;

exports[`customers_subscriptions .cancel() should return a subscription instance 1`] = `
Subscription {
  "_links": Object {
    "customer": Object {
      "href": "https://api.mollie.com/v2/customers/cst_stTC2WHAuS",
      "type": "application/hal+json",
    },
    "self": Object {
      "href": "https://api.mollie.com/v2/customers/cst_stTC2WHAuS/subscriptions/sub_rVKGtNd6s3",
      "type": "application/hal+json",
    },
  },
  "amount": Object {
    "currency": "EUR",
    "value": "25.00",
  },
  "canceledAt": null,
  "createdAt": "2018-06-01T12:23:34+00:00",
  "description": "Quarterly payment",
  "id": "sub_rVKGtNd6s3",
  "interval": "3 months",
  "method": null,
  "mode": "live",
  "resource": "subscription",
  "startDate": null,
  "status": "active",
  "times": 4,
  "webhookUrl": "https://webshop.example.org/subscriptions/webhook",
}
`;

exports[`customers_subscriptions .cancel() should work with a callback 1`] = `
Subscription {
  "_links": Object {
    "customer": Object {
      "href": "https://api.mollie.com/v2/customers/cst_stTC2WHAuS",
      "type": "application/hal+json",
    },
    "self": Object {
      "href": "https://api.mollie.com/v2/customers/cst_stTC2WHAuS/subscriptions/sub_rVKGtNd6s3",
      "type": "application/hal+json",
    },
  },
  "amount": Object {
    "currency": "EUR",
    "value": "25.00",
  },
  "canceledAt": null,
  "createdAt": "2018-06-01T12:23:34+00:00",
  "description": "Quarterly payment",
  "id": "sub_rVKGtNd6s3",
  "interval": "3 months",
  "method": null,
  "mode": "live",
  "resource": "subscription",
  "startDate": null,
  "status": "active",
  "times": 4,
  "webhookUrl": "https://webshop.example.org/subscriptions/webhook",
}
`;

exports[`customers_subscriptions .create() should return a subscription instance 1`] = `
Subscription {
  "_links": Object {
    "customer": Object {
      "href": "https://api.mollie.com/v2/customers/cst_stTC2WHAuS",
      "type": "application/hal+json",
    },
    "self": Object {
      "href": "https://api.mollie.com/v2/customers/cst_stTC2WHAuS/subscriptions/sub_rVKGtNd6s3",
      "type": "application/hal+json",
    },
  },
  "amount": Object {
    "currency": "EUR",
    "value": "25.00",
  },
  "canceledAt": null,
  "createdAt": "2018-06-01T12:23:34+00:00",
  "description": "Quarterly payment",
  "id": "sub_rVKGtNd6s3",
  "interval": "3 months",
  "method": null,
  "mode": "live",
  "resource": "subscription",
  "startDate": null,
  "status": "active",
  "times": 4,
  "webhookUrl": "https://webshop.example.org/subscriptions/webhook",
}
`;

exports[`customers_subscriptions .get() should return a subscription instance 1`] = `
Subscription {
  "_links": Object {
    "customer": Object {
      "href": "https://api.mollie.com/v2/customers/cst_stTC2WHAuS",
      "type": "application/hal+json",
    },
    "self": Object {
      "href": "https://api.mollie.com/v2/customers/cst_stTC2WHAuS/subscriptions/sub_rVKGtNd6s3",
      "type": "application/hal+json",
    },
  },
  "amount": Object {
    "currency": "EUR",
    "value": "25.00",
  },
  "canceledAt": null,
  "createdAt": "2018-06-01T12:23:34+00:00",
  "description": "Quarterly payment",
  "id": "sub_rVKGtNd6s3",
  "interval": "3 months",
  "method": null,
  "mode": "live",
  "resource": "subscription",
  "startDate": null,
  "status": "active",
  "times": 4,
  "webhookUrl": "https://webshop.example.org/subscriptions/webhook",
}
`;
